UNIXFS.md

```diff
@@ -59,6 +64,61 @@ For files that are comprised of more than a single block, the 'Type' field will
 
 This data is serialized and placed inside the 'Data' field of the outer merkledag protobuf, which also contains the actual links to the child nodes of this object.
 
+## Metadata
+
+UnixFS currently supports three metadata fields:
+
+* `mimeType` -- The mime-type of the file. This generally shouldn't be used.
+* `mode` -- The `mode` is for optionally persisting the [file permissions in numeric notation](https://en.wikipedia.org/wiki/File_system_permissions#Numeric_notation) \[[spec](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_stat.h.html)\]. It defaults to 0755 if unspecified.
+* `mtime` -- The modification time in seconds since the epoch. This defaults to the unix epoch if unspecified.
+
+There are three ways to specify file metadata:
+
+### Embedded in the directory
+
+Each entry in the repeated metadata field corresponds to the linked file/directory at the same offset in the Links section of the outer dag-pb. This field is appropriate in nodes with type `Directory` and `HAMTShard`. However, metadata should _not_ be specified for links that point to other HAMT shards.
+
+For some context, this solution was chosen over embedding the metadata in the file itself because:
+
+1. It allows accessing the metadata without downloading the target files.
+2. More importantly, it avoids changing the root hash of the target files to allow for better deduplication.
```

Comment:
Should we mention the tradeoff? If I have `QmDir` and `file.txt` (`QmFileTxt`) then I'll only be able to see metadata for the file if it's accessed via `/ipfs/QmDir/file.txt`. The file will have no metadata when accessed like `/ipfs/QmFileTxt`.

Reply:
@alanshaw from https://github.com/ipfs/specs/pull/220#issuecomment-530028458 :

> As far as I can tell, for the original set of users waiting on this, there is no usecase where a piece of data needs POSIX attributes without a filename. The filename is already only available as a directory wrapper node, there is no other way to attach name to an "anonymous stream of data". What makes other attributes different enough to warrant a non-directory anonymous wrapper-node?

I.e. what is the use-case of needing to know mtime/nodetype/permissions of an object **without** knowing its name?

Reply:
In this design, the metadata is stored in the parent directory. That doesn't mean that it's tied to the file name.

It _does_ mean that simplifying a path from `/ipfs/QmDir/file.txt` to `/ipfs/QmFile` _will_ drop the metadata (along with the filename). The tradeoff is:

1. Better deduplication.
2. Efficient file listing (keeping the metadata in the directory means we can list a directory, including things like modes, without resolving the actual objects).

Reply:
This doesn't seem like the right way to do it to me.  If I copy a file to another location on my filesystem, it doesn't lose its permissions, nor will it lose its permissions if I remove it from its containing directory (to put it in a tarball, for example).

It will involve changing the `CID` of a file if it's metadata changes but that seems like a reasonable tradeoff to me, since the `CID`s of the underlying file chunks will not change (assuming the file size is > 256KB).  We'll probably see the most `CID` churn in a users' `mfs`, but we generally don't care about `CID`s in `mfs` as much as paths.

Reply:
A counter thought: why should metadata that changes independently from the data (e.g. permissions) be anywhere in the directory tree, couldn't it live outside the tree? 

> If I copy a file to another location on my filesystem, it doesn't lose its permissions

True, but when I email you a file or upload it to Dropbox it does lose its permissions. Similarly, using unix time as a metadata attribute makes very little sense in a distributed context. For example, Alice and Bob both download an Ubuntu ISO and add it to IPFS the timestamps will be different.

> It will involve changing the CID of a file if it's metadata changes but that seems like a reasonable tradeoff to me, since the CIDs of the underlying file chunks will not change

If I recall correctly, the way IPFS currently works makes this a pretty big performance hit. For example, say Alice and Bob both have the same files, but with different metadata. Then if Charlie tries to download Alice's CID while Alice is offline then he gets no data. Similarly, even if Alice is online Charlie will end up downloading the data from Alice but not Bob since his node probably won't make another DHT request for the file CID (which would be expensive anyway).

There are alternative solutions for people that want to share metadata that do not involve putting OS metadata near files for everyone else, which I think are worth discussing. However, if unixfsv1.5 is just a temporary measure to tide us over until unixfsv2 lands then as long as we only use the metadata we actually need (i.e. mtime + execute bit) then this is fine by me.

Reply:
> when I email you a file or upload it to Dropbox it does lose its permissions.

Those aren't filesystem operations so I would not expect them to retain permissions.  Ok, tarballing something up isn't specifically a filesystem operation either, but it is sensitive to it's environment so does preserve perms, I think we should endeavour to do the same thing.

> Similarly, using unix time as a metadata attribute makes very little sense in a distributed context.

I agree, distributed time is a bit weird.  That said, the file chunks of the ISO will be the same, just the root node containing the metadata will be different.  *That* said Alice and Bob could use different chunking strategies, or whatever so it's already easy to have two CIDs that point to the same file.

> the way IPFS currently works

This doesn't seem like a good thing to base spec decisions on.  Changing how IPFS works is just another spec decision away, after all..

Reply:
> Those aren't filesystem operations so I would not expect them to retain permissions.

I think that people having mounted Dropbox folders on their machine almost exactly mimics the behavior that users would expect from IPFS.

> could use different chunking strategies, or whatever so it's already easy to have two CIDs that point to the same file.

That's true, but A) if they the chunking defaults there's no problem B) there are possible solutions for dealing with multiple chunkers since fundamentally the two files can have a canonical CID (e.g. sha256 of the entire file without any unixfs intermediate blocks), however these solutions cannot properly work if the actual content is different (which it would be since metadata contains information)

> Changing how IPFS works is just another spec decision away, after all

Unless there's a proposal for how to fix the problem I'm wary of "we'll fix it later". Also, my understanding is that unixfsv1.5 is supposed to be a temporary fix and so the probability of "later" coming before unixfsv1.5 is deprecated is even lower.

Reply:
> This doesn't seem like the right way to do it to me. If I copy a file to another location on my filesystem, it doesn't lose its permissions, nor will it lose its permissions if I remove it from its containing directory (to put it in a tarball, for example).

@achingbrain re-reading the thread I realized this is also a flawed model. In general it is the tool that does the copying that preserves various bits of metadata, and moreover only does so **if the context allows it to**. Consider:
```
rabbit@Ahasver:/dev/shm/stuff$ ls -l
total 0

rabbit@Ahasver:/dev/shm/stuff$ touch source

rabbit@Ahasver:/dev/shm/stuff$ chmod 666 source

rabbit@Ahasver:/dev/shm/stuff$ umask
0022

rabbit@Ahasver:/dev/shm/stuff$ cp source dest

rabbit@Ahasver:/dev/shm/stuff$ ls -l
total 0
-rw-r--r-- 1 rabbit rabbit 0 Nov 13 10:29 dest
-rw-rw-rw- 1 rabbit rabbit 0 Nov 13 10:29 source
```

Reply:
The default behaviour is to create the new file with the same permissions, `umask` changes the defaults.

If we access a file via it's CID we won't have the option to preserve permissions as they are simply not there.